src/core/matcher.rkt

```diff
@@ -64,11 +64,29 @@
 
 (define (rewrite-once expr ctx #:rules rules #:root [root-loc '()])
   (define expr-repr (repr-of expr ctx))
-  (reap [sow]
-    (for ([rule rules] #:when (equal? expr-repr (rule-otype rule)))
-      (let* ([result (rule-apply rule expr)])
-        (when result
-          (sow (list (change rule root-loc (cdr result)))))))))
+  (define (match-otype? otype)
+    (or (equal? otype (representation-type expr-repr))
+        (equal? otype expr-repr)))
```

Comment:
Hmm, I'm a bit confused. `match-otype?` is only called on the result of `rule-otype`; why can that be either a type or a representation?

Reply:
This should change. Since the rules are expanded it should be only representation